File: 2025_3/lc_3318.py
from collections import Counter
class Solution:
    def findXSum(self, nums: List[int], k: int, x: int) -> List[int]:
        #배열에서 요소 발생 횟수 카운트
        #top x 만큼 발생하는 것만. 둘이 같은 빈도면 더 큰 값 선택
        #결과 배열의 sum 계산
        #x만큼 구별된 요소 없으면 xsum 은 배열의 전체합
        num = Counter(nums[:k])
        sorted_num = sorted(num.items(), key=lambda x: (-x[1], -x[0]))
        ans =[]
        ans.append(sum(num * cnt for num, cnt in sorted_num[:x]))

        for i in range(k, len(nums)):
            old = nums[i-k]
            new = nums[i] 

            if old in num:
                num[old] -= 1
                if num[old] == 0:
                    del num[old]
            num[new] += 1

          
            sorted_num = sorted(num.items(), key=lambda x: (-x[1], -x[0]))

            ans.append(sum(num * cnt for num, cnt in sorted_num[:x]))

        return ans

File: 2025_3/lc_3318.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} x
 * @return {number[]}
 */
function getTopX(hash, x) {
    let sorted = Object.entries(hash)
        .map(([key, value]) => [parseInt(key), value])    
        .sort((a,b) => b[1] - a[1] || b[0] - a[0])
    return sorted.slice(0, x).reduce((acc, entry) => acc + entry[0] * entry[1], 0);
    }
var findXSum = function(nums, k, x) {
    //top x 의 합계를 구하는 함수
    
    let hash = {}
    let ans = []
    let k_num = nums.slice(0,k)
    console.log(k_num)

    for(let num of k_num) {
        hash[num] = (hash[num] || 0) + 1;
    }

    ans.push(getTopX(hash, x));

    for (let j=k; j<nums.length; j++) {
        let oldNum = nums[j-k]
        let newNum = nums[j]

        hash[oldNum] -= 1
        if (hash[oldNum] == 0) {
            delete hash[oldNum]
        }

        hash[newNum] = (hash[newNum] || 0) + 1;
        ans.push(getTopX(hash, x));
    }

    return ans
  
};